exec/java-exec/src/main/java/org/apache/drill/exec/planner/fragment/ZKQueueParallelizer.java

```diff
@@ -0,0 +1,120 @@
+import java.util.Collection;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+public class ZKQueueParallelizer extends SimpleParallelizer {
+
+  private final boolean planHasMemory;
+  private final QueryContext queryContext;
+  private Map<String, Collection<PhysicalOperator>> endpointMap;
+  private final QueryResourceManager resourceManager;
+
+  public ZKQueueParallelizer(boolean memoryAvailableInPlan, QueryResourceManager rm, QueryContext queryContext) {
+    super(queryContext);
+    this.planHasMemory = memoryAvailableInPlan;
+    this.queryContext = queryContext;
+    this.resourceManager = rm;
+  }
+
+  @Override
+  public void adjustMemory(PlanningSet planningSet, Set<Wrapper> roots,
+                           Map<DrillbitEndpoint, String> onlineEndpointUUIDs) throws PhysicalOperatorSetupException {
+    if (planHasMemory) {
+      return;
+    }
+
+    Collector collector = new Collector();
+
+    for (Wrapper wrapper : roots) {
+      traverse(wrapper, CheckedConsumer.throwingConsumerWrapper((Wrapper fragment) -> {
+        fragment.getNode().getRoot().accept(collector, fragment);
+      }));
+    }
+
+    endpointMap = collector.getNodeMap();
+    ZKQueueMemoryAllocationUtilities.planMemory(queryContext, this.resourceManager, endpointMap);
+  }
+
+
+  public class Collector extends AbstractOpWrapperVisitor<Void, RuntimeException> {
+    private final Multimap<DrillbitEndpoint, PhysicalOperator> bufferedOperators;
+
+    public Collector() {
+      this.bufferedOperators = ArrayListMultimap.create();
+    }
+
+    private void getMinorFragCountPerDrillbit(Wrapper currFragment, PhysicalOperator operator) {
+      for (DrillbitEndpoint endpoint : currFragment.getAssignedEndpoints()) {
+        bufferedOperators.put(endpoint, operator);
+      }
+    }
+
+    @Override
+    public Void visitSendingExchange(Exchange exchange, Wrapper fragment) throws RuntimeException {
+      return visitOp(exchange, fragment);
+    }
+
+    @Override
+    public Void visitReceivingExchange(Exchange exchange, Wrapper fragment) throws RuntimeException {
+      return null;
+    }
+
+    @Override
+    public Void visitOp(PhysicalOperator op, Wrapper fragment) {
+      if (op.isBufferedOperator(queryContext)) {
+        getMinorFragCountPerDrillbit(fragment, op);
+      }
+      for (PhysicalOperator child : op) {
+        child.accept(this, fragment);
+      }
+      return null;
+    }
+
+    public Map<String, Collection<PhysicalOperator>> getNodeMap() {
+      Map<DrillbitEndpoint, Collection<PhysicalOperator>> endpointCollectionMap = bufferedOperators.asMap();
+      Map<String, Collection<PhysicalOperator>> nodeMap = new HashMap<>();
+      for (Map.Entry<DrillbitEndpoint, Collection<PhysicalOperator>> entry : endpointCollectionMap.entrySet()) {
+        nodeMap.put(entry.getKey().getAddress(), entry.getValue());
+      }
+
+      return nodeMap;
+    }
+  }
```

Comment:
Couple of things:
1) Why not construct `Map<String, Collection<PhysicalOperator>>` to start with instead of `Multimap<DrillbitEndpoint, PhysicalOperator> `and then converting it.
2) `bufferedOperators.asMap()` is not guaranteed to work in the same way as intended since DrillbitEndpoint `equals` and `hashCode` methods are not reliable which is why we created `DrillNode`.